File: Dataset/Leetcode/train/100/165.cpp
 class Solution {
public:
    bool XXX(TreeNode* p, TreeNode* q) {
        stack<TreeNode*>sp,sq;
        while(true)
        {
            while(p!=NULL||q!=NULL)
            {
                if(p==NULL||q==NULL||p->val!=q->val)
                    return false;
                sp.push(p);
                sq.push(q);
                p=p->left;
                q=p->left;
            }
            if(sp.empty()&&sq.empty()) break;
            p=sp.top();
            sp.pop();
            p=p->right;
            q=sq.top();
            sq.pop();
            q=q->right;
        }
        return true;
    }
};

File: Dataset/Leetcode/test/110/165.java
class Solution {
    public boolean XXX(TreeNode root) {
        if(root==null) return true;//空树
        if(Math.abs(depth(root.left)-depth(root.right))>1) return false; //左右子树高度差不大于1
        return XXX(root.left)&&XXX(root.right);//左右子树也是搜索二叉树
    }
    public int depth(TreeNode root){
        if(root==null) return 0;
        return Math.max(depth(root.left),depth(root.right))+1;
    }
}

